### DISCUSSION TITLE
WebView2 on Xbox with GDK

### DISCUSSION THREAD

**Discussion by AshleyScirra** [MicrosoftEdge/WebView2Feedback#4149] [Q&A]
I'm finding it difficult to understand all the technologies involved (which seem to have changed several times over the years), but as far as I can tell, the GDK is now the preferred way to publish to Xbox ([ref](https://github.com/microsoft/GDK)):

> Using Win32 + GDK is the primary, supported app model to build games for Xbox console, Xbox Game Pass (both Xbox and PC), and Xbox Game Streaming**.

Currently it seems WebView2 on Xbox is only supported via UWP, which apparently is no longer the preferred way to develop for Xbox:

> UWP apps and games are community-supported only; partners inside Xbox managed programs (Xbox, Xbox Game Pass, Xbox Game Streaming) should use Microsoft Win32 + GDK.

So it seems that while there is WebView2/Windows/Win32, WebView2/Windows/UWP, and WebView2/Xbox/UWP, there isn't yet WebView2/Xbox/Win32, which precludes using the GDK.

Have I got that right? Is support for WebView2 on Xbox with Win32 + GDK planned? This seems to be important for publishing web games to Xbox using WebView2.

**Comment by victorhuangwq** [comment]
@nishitha-burman could you help @AshleyScirra out?

**Comment by AshleyScirra** [comment]
Perhaps we can use the GDK via Win32 calls inside a UWP app? [This article](https://learn.microsoft.com/en-us/uwp/win32-and-com/win32-and-com-for-uwp-apps) appears to indicate it's supported to some extent, but it's not clear if this is supported/permitted on Xbox.

**Comment by la-thorn** [comment]
In general, the GDK is the recommended framework for shipping (traditional) games on Xbox. But if you're planning to build a web-based game, it may make more sense to build it on top of UWP so you can take advantage of WebView2. UWP is still the recommended platform for building *apps* for Xbox. The GDK does not offer any kind of WebView2 integration.

**Comment by AshleyScirra** [comment]
Yes, I think using WebView2 means we have to use UWP. But do you know if we can use the GDK from C++ code in a UWP app? I'd happily write all the integration myself if that's allowed. I guess I should ask someone on the Xbox team too, and I'm trying to get in touch with someone there.

**Reply by victorhuangwq** [reply]
@nthorn should be able to give you a clear answer, as if I'm not wrong, they are from the Xbox team.

**Reply by la-thorn** [reply]
It is not possible to call GDK APIs from a UWP app, the two platforms are mutually exclusive. (And yep, I'm on the Xbox team!)

**Reply by AshleyScirra** [reply]
OK, that's a shame... I think it seriously limits publishing options for web games. The Creator's Program supports UWP games, but it doesn't look well maintained (the official libraries for integration are now ~5 years without update). If we want a webview on Xbox with access to the GDK then I guess we can't use WebView2, right? Does it still support the legacy WebView though? Is there any scope for modern WebView2 to be supported with Win32 on Xbox?

**Reply by la-thorn** [reply]
There are no current plans to bring WebView2 to the GDK; legacy WebView is not supported on top of the GDK either. If you're looking to display web content in a GDK application your best bet is to look for a 3rd party middleware solution.

**Reply by AshleyScirra** [reply]
That's a real shame - for us one of the main points of WebView2 is to avoid these third-party custom-compiled browser engines like Electron and NW.js. In particular it also risks losing JIT support and so a lot of performance if the code generation is not allowed to be done in the app instead of by the system (although I don't know what the Xbox rules around this are exactly).

I'm not aware that Electron or NW.js have any Xbox ports, and a web search has turned up nothing for middleware that provides a browser engine that it says works on Xbox. Is that all behind NDA? Can you point me in the direction of any public middleware providers that have a browser engine that can run on Xbox with the GDK?

**Reply by la-thorn** [reply]
These are the two that I'm personally aware of:
* https://coherent-labs.com/
* https://ultralig.ht/

Though these are more commonly used for game UI; I don't know if they would be effective for rendering your whole application.

**Reply by AshleyScirra** [reply]
OK, thanks for the pointers. It's not all good news for us but thanks for the help and taking the time to answer my questions, I appreciate it.

**Reply by AshleyScirra** [reply]
(For what it's worth: Coherent Gameface is [not a full browser engine](https://docs.coherent-labs.com/cpp-gameface/what_is_gfp/overview/), and Ultralight [lacks key features like WebGL and HTML5 audio](https://github.com/ultralight-ux/Ultralight/issues/178), so I don't think either are feasible. I do think there's a gap in the market for a full Chromium browser engine for the GDK, but we don't have the resources to do that.)
